ecosystem/indexer-grpc/indexer-transaction-generator/src/test_case.rs

```diff
@@ -0,0 +1,263 @@
+/// Each test case is a folder containing:
+///  - folder: setup_1, setup_2, ...: the setup __folders__ to prepare the test environment.
+///  - action_1, action_2, ...: the action __folders__ to execute the test.
+///  - test_config.json: the test case configuration file.
+/// Test case is expected to be run in order: setups first, then actions.
+pub struct TestCase {
+    // Test case name.
+    pub name: String,
+    // Ordered steps from setup to action.
+    pub steps: Vec<Step>,
+    // Test case configuration.
+    pub test_config: TestCaseConfig,
+}
+
+impl TestCase {
+    pub fn load(test_case_path: PathBuf) -> Result<Self, anyhow::Error> {
+        // Load the test case configuration.
+        let test_config_path = test_case_path.join(TEST_CONFIG_FILE);
+        let test_config_raw_string = std::fs::read_to_string(test_config_path)
+            .context("Failed to read test config file.")?;
+        let test_config: TestCaseConfig = serde_json::from_str(&test_config_raw_string)
+            .context("Failed to parse test config file.")?;
+
+        // Get steps.
+        let mut steps = vec![];
+        let test_case = std::fs::read_dir(&test_case_path).context(format!(
+            "Failed to read test case folder at {:?}",
+            &test_case_path
+        ))?;
+        for step in test_case {
+            let step = step.context(format!(
+                "Failed to read step folder at {:?}",
+                &test_case_path
+            ))?;
+            let step_path = step.path();
+            let step = Step::try_from(step_path)?;
+            steps.push(step);
+        }
+
+        // Filter out the unknown steps.
+        let mut steps: Vec<Step> = steps
+            .into_iter()
+            .filter(|step| !matches!(step, Step::UNKNOWN))
+            .collect();
+
+        // Sort the steps by priority.
+        steps.sort_by_key(|step| step.priority());
+
+        let name = test_case_path
+            .file_name()
+            .context(format!(
+                "Failed to get test case name at {:?}",
+                &test_case_path
+            ))?
+            .to_str()
+            .context(format!(
+                "Failed to get test case name at {:?}",
+                &test_case_path
+            ))?
+            .to_string();
+        Ok(TestCase {
+            name,
+            steps,
+            test_config,
+        })
+    }
+}
+
+/// Step is a wrapper around the path of the setup and action folders.
+/// It's to facilitate the validation of the test case.
+/// Error vs UNKNOWN:
+///  - If a PathBuf can be regex matched to a setup or action step, but cannot be parsed as a folder, it's an error.
+///  - If a PathBuf cannot be regex matched to a setup or action folder, it's UNKNOWN.
+#[derive(PartialEq, Eq, Debug, Hash)]
+pub enum Step {
+    Setup((PathBuf, u64)),
+    Action((PathBuf, u64)),
+    /// This is to handle the case when the step is not a setup or action folder.
+    UNKNOWN,
```

Comment:
Out of curiosity when do we expect additional unknown folders like this? Might be good to explain in the comments.

Reply:
done done done.

like `README.md` or `.ds_store` to ignore.